### ISSUE TITLE
Add support of QOI images

### ISSUE THREAD

**Issue by caffeine-mgn** [godotengine/godot-proposals#3726]
### Describe the project you are working on

Making some game

### Describe the problem or limitation you are having in your project

I want to use QOI images for fast loading. More info this: https://github.com/phoboslab/qoi

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Faster image loading

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

N/A

### If this enhancement will not be used often, can it be worked around with a few lines of script?

nop

### Is there a reason why this should be core and not an add-on in the asset library?

Yep, image decoding looks like a part of core

**Comment by Calinou** [comment]
> I want to use QOI images for fast loading. More info this: phoboslab/qoi

Have you profile image loading to be a bottleneck in your project? [Lossless WebP encoding](https://github.com/godotengine/godot/pull/47835) was added to Godot 3.4, and it already loads faster than PNG did.

**Comment by fire** [comment]
See also https://github.com/godotengine/godot-proposals/issues/2461

**Comment by IntangibleMatter** [comment]
I think that adding QOI would be a good thing. It's pretty small, fast, and is getting less and less niche. I know that as soon as I can I'm moving away from PNG. If nothing else, QOI seems "cooler" to devs and Godot having native support would be a good thing to do for the engine in the long term. Be a leader.

Besides, qoi.h is only like, 600 lines long :p

**Comment by Calinou** [comment]
> If nothing else, QOI seems "cooler" to devs and Godot having native support would be a good thing to do for the engine in the long term. Be a leader.

I'm afraid this is not how decisions are taken in Godot. See [Best practices for engine contributors](https://docs.godotengine.org/en/stable/community/contributing/best_practices_for_engine_contributors.html) for more information.

**Comment by DeeJayLSP** [comment]
After finding out that GameMaker has an implementation of QOI (combined with BZip2) as the default image compression I decided to run a few tests.

Since Godot compresses using Lossless WebP by default, the comparison would be against it at max compression. While QOI undoubtedly encodes much faster (and likely has a close decoding speed), it loses almost every time in compression ratio, even to PNG.

The situation changes if you include BZip2 compression. While it now always beats PNG, it *sometimes* beats Lossless WebP by being 10-23% smaller, sometimes loses by being 10-38% bigger.

This is just something I did out of curiosity and wanted to share since it's in the context of this proposal.

**Comment by fire** [comment]
Since we already use zstd, maybe there's a setting of zstd with QOI that makes it enter the competition.

**Comment by Calinou** [comment]
Compressed QOI is an interesting idea, but it makes the file format no longer standard. This means ffmpeg can't accept it as input for Movie Maker usage, for instance.

**Comment by DeeJayLSP** [comment]
My consideration was Compressed QOI as a potential alternative to Lossless WebP compression. But as I said, WebP still wins.

**Comment by knuxyl** [comment]
I'm suprised this hasn't been done yet. The format is one of the simplest and this would help old/bad hardware by freeing up cpu for other stuff. QOI isn't about storage space, it's encode/decode speed.

**Comment by Calinou** [comment]
> this would help old/bad hardware by freeing up cpu for other stuff

VRAM-compressed textures (i.e. most textures in 3D games) don't rely on traditional image formats like PNG or QOI, so this wouldn't affect their loading times. In 2D, Godot uses lossless WebP by default which is already pretty fast to decode (faster than PNG).

**Comment by knuxyl** [comment]
> > this would help old/bad hardware by freeing up cpu for other stuff
> 
> VRAM-compressed textures (i.e. most textures in 3D games) don't rely on traditional image formats like PNG or QOI, so this wouldn't affect their loading times. In 2D, Godot uses lossless WebP by default which is already pretty fast to decode (faster than PNG). See [@DeeJayLSP](https://github.com/DeeJayLSP)'s comment above.

well i care mainly about the editor. i do a lot of batch image conversions and alterations with imagemagick and qoi is much much faster and itd be great if i didnt have to go and convert it all to png. does godot use gpu to render 2d graphics? ie if i made a card game, does the gpu or image format matter most? does godot convert all sprite assets to webp during "compilation"?

**Comment by fire** [comment]
> Describe the problem or limitation you are having in your project

My main issue with this proposal is that QOI is a way to have faster encoding and decoding for processing, but it's a lot of work for something most people don't use.

We could replace the internal WEBP method with QOI, which should have some benefits. Does QOI have a lossless mode?

Thoughts?

**Comment by knuxyl** [comment]
> > Describe the problem or limitation you are having in your project
> 
> My main issue with this proposal is that QOI is a way to have faster encoding and decoding for processing, but it's a lot of work for something most people don't use.
> 
> We could replace the internal WEBP method with QOI, which should have some benefits. Does QOI have a lossless mode?
> 
> Thoughts?
> 
> Edited:
> 
> I'll support QOI if someone can demonstrate replacing the internal WEBP method with QOI and show the benefits.

qoi is only lossless. the algorithm for qoi is quite simple, it shouldn't be too much work for someone who knows the structure of godots source.

**Comment by fire** [comment]
@DeeJayLSP, did you try compressing QOI inside a RES file using ZSTD compression? What numbers did you get?

https://docs.godotengine.org/en/stable/classes/class_resourcesaver.html#class-resourcesaver-method-save

**Comment by DeeJayLSP** [comment]
> @DeeJayLSP, did you try compressing QOI inside a RES file using ZSTD compression? What numbers did you get?
> 
> https://docs.godotengine.org/en/stable/classes/class_resourcesaver.html#class-resourcesaver-method-save

What I tried was compressing a QOI file using the `zst` command, using the maximum compression settings I could find. 

Not sure if it's the same thing, but in most cases lossless WebP still won.

**Comment by fire** [comment]
I don't think we should add QOI unless we can use it by default for internal compression. It was not a winner in compression size (unless there's other evidence), therefore I can't support adding it.

QOI is an optimization for developer iteration speed as mentioned by the proposal, but it is niche and most people don't use it. 

What are the current winners for developer iteration out of the current Godot Engine formats?

Curious.

Maybe we can improve developer iteration speed on read, update and write cycles.

**Comment by DeeJayLSP** [comment]
I mentioned the case of GameMaker once, where it uses QOI combined with BZip2 for lossless compression. I did a test of with a bunch of images (convert to QOI, compress using `bzip2` command) and while the margin compared to Lossless WebP was lower, it still lost in most cases. The compression format that got the closest to Lossless WebP was Brotli, but even then lost many times.

The **only** benchmark where QOI absolutely wins over most lossless compressed formats is compression speed. And I couldn't find a use case for this except for Movie Writer frames (FFmpeg supports QOI, so would be a good interchange format).

**Comment by DeanLemans** [comment]
Qoi wont be added as decided by technical comitee, shouldnt this be closed then?

**Comment by DeeJayLSP** [comment]
The only considerable advantage of QOI over other lossless-compressed formats is its blazing fast encoding and decoding speed. But compare it with Lossless WebP: while encoding is sluggish (unless you use a low level, which makes it in pair with PNG), its decoding speed is in pair with QOI, and the compression ratio is ***much*** higher, even at lower levels.

If saving QOI (mainly for MovieWriter purposes due to its encoding speed and the fact FFmpeg supports QOI) was judged unsuitable, I don't see a reason for adding support for loading/importing it either. It's not like we'll be finding and downloading assets distributed as QOI.

The [article announcing QOI](https://phoboslab.org/log/2021/11/qoi-fast-lossless-image-compression) clearly mentions it was meant to compete with PNG, but when much newer formats that better serve the purpose exist, why bother with it?
